### PR TITLE
typing: add gateway event types

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Adds `TypedDict`s for all gateway events, and fixes some minor issues found through that, like thread list sync missing an `int` cast.

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/types/gateway.py [hunk 1]
```diff
@@ -241,3 +257,352 @@ class VoiceResumeData(TypedDict):
 class VoiceResumeCommand(TypedDict):
     op: Literal[7]
     d: VoiceResumeData
+
+
+# events
```
Comment (Bluenix2) [inline disnake/types/gateway.py:262]: This comment looks a little awkward 😄

**Review by Bluenix2 — COMMENTED** [review]
Other than my one comment, this looks good!

File: disnake/types/gateway.py (hunk 1 above)
Comment (shiftinv) [reply disnake/types/gateway.py]: Hmm, good point 😅 Tried to improve this a bit in 470156227e22ba706e48479fcd9bdaf5d3fb13f6.

File: disnake/state.py [hunk 2]
```diff
@@ -912,11 +922,11 @@ def parse_interaction_create(self, data) -> None:
             elif interaction.data.component_type is ComponentType.select:
                 self.dispatch("dropdown", interaction)
 
-        elif interaction_type == 4:
+        elif data["type"] == 4:
             interaction = ApplicationCommandInteraction(data=data, state=self)
             self.dispatch("application_command_autocomplete", interaction)
 
-        elif interaction_type == 5:
+        elif data["type"] == 5:
```
Comment (onerandomusername) [inline disnake/state.py:929]: Could we please revert these changes?

Comment (shiftinv) [reply disnake/state.py:929]: Doing it this way, without an intermediate variable, helps pyright narrow the data union to the correct type - otherwise we'd have to `# type: ignore` all the `*Interaction(data=data, state=self)` calls, which I'd really like to avoid.
The performance overhead is negligible, with a difference of `(10.9 us - 8.65 us) / 100 = 22.5 ns` per call: https://workbin.dev/?id=1658072299515132

File: disnake/state.py [hunk 3]
```diff
@@ -875,10 +881,16 @@ def parse_message_reaction_remove(self, data) -> None:
...
+    def parse_message_reaction_remove_emoji(
+        self, data: gateway.MessageReactionRemoveEmojiEvent
+    ) -> None:
         emoji = data["emoji"]
         emoji_id = utils._get_as_snowflake(emoji, "id")
-        emoji = PartialEmoji.with_state(self, id=emoji_id, name=emoji["name"])
+        emoji = PartialEmoji.with_state(
+            self,
+            id=emoji_id,
+            name=emoji["name"],  # type: ignore
```
Comment (onerandomusername) [inline disnake/state.py:892]: please leave a comment why

File: disnake/state.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/state.py:929]: That's unfortunate, I suppose. Should probably have a comment explaining why

File: disnake/state.py [hunk 4]
```diff
@@ -1308,19 +1322,18 @@ def parse_guild_stickers_update(self, data) -> None:
...
+    def _get_create_guild(self, data: gateway.GuildCreateEvent) -> Guild:
         if data.get("unavailable") is False:
             # GUILD_CREATE with unavailable in the response
             # usually means that the guild has become available
             # and is therefore in the cache
             guild = self._get_guild(int(data["id"]))
             if guild is not None:
                 guild.unavailable = False
-                guild._from_data(data)
+                guild._from_data(data)  # type: ignore
```
Comment (onerandomusername) [inline disnake/state.py:1336]: explain why

File: disnake/state.py [hunk 5]
```diff
@@ -856,10 +858,14 @@ def parse_message_reaction_remove_all(self, data) -> None:
...
 
-    def parse_message_reaction_remove(self, data) -> None:
+    def parse_message_reaction_remove(self, data: gateway.MessageReactionRemoveEvent) -> None:
         emoji = data["emoji"]
         emoji_id = utils._get_as_snowflake(emoji, "id")
-        emoji = PartialEmoji.with_state(self, id=emoji_id, name=emoji["name"])
+        emoji = PartialEmoji.with_state(
+            self,
+            id=emoji_id,
+            name=emoji["name"],  # type: ignore
```
Comment (onerandomusername) [inline disnake/state.py:867]: comment why

File: disnake/invite.py [hunk 6]
```diff
@@ -511,7 +509,12 @@ def from_gateway(cls, *, state: ConnectionState, data: GatewayInvitePayload) ->
             guild = Object(id=guild_id) if guild_id is not None else None
             channel = Object(id=channel_id)
 
-        return cls(state=state, data=data, guild=guild, channel=channel)  # type: ignore
+        return cls(
+            state=state,
+            data=data,
+            guild=guild,  # type: ignore
+            channel=channel,  # type: ignore
```
Comment (onerandomusername) [inline disnake/invite.py:516]: comments why

File: disnake/state.py [hunk 7]
```diff
@@ -587,8 +582,11 @@ def _get_message(self, msg_id: Optional[int]) -> Optional[Message]:
             else None
         )
 
-    def _add_guild_from_data(self, data: GuildPayload) -> Guild:
-        guild = Guild(data=data, state=self)
+    def _add_guild_from_data(self, data: Union[GuildPayload, UnavailableGuildPayload]) -> Guild:
+        guild = Guild(
+            data=data,  # type: ignore
```
Comment (onerandomusername) [inline disnake/state.py:587]: comment why

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/state.py (hunk 2 above)
Comment (shiftinv) [reply disnake/state.py:945]: https://github.com/DisnakeDev/disnake/pull/594/commits/fc2be126a25524eda99b3018e8124c48d2574f31

File: disnake/state.py (hunk 3 above)
Comment (shiftinv) [reply disnake/state.py:902]: https://github.com/DisnakeDev/disnake/pull/594/commits/fc2be126a25524eda99b3018e8124c48d2574f31
This should be improved in a future PR, but it's out of scope for this one.

File: disnake/state.py (hunk 4 above)
Comment (shiftinv) [reply disnake/state.py]: https://github.com/DisnakeDev/disnake/pull/594/commits/fc2be126a25524eda99b3018e8124c48d2574f31

File: disnake/invite.py (hunk 6 above)
Comment (shiftinv) [reply disnake/invite.py:517]: https://github.com/DisnakeDev/disnake/pull/594/commits/fc2be126a25524eda99b3018e8124c48d2574f31

**Review by onerandomusername — APPROVED** [review]
lgtm, lets do it
